gcloud/bigquery/table.py

```diff
@@ -693,6 +699,225 @@ def insert_data(self,
+                                ignore_unknown_values, max_bad_records,
+                                quote_character, skip_leading_rows,
+                                write_disposition)
+
+        upload = Upload(file_obj, content_type, total_bytes,
+                        auto_transfer=False)
+
+        url_builder = _UrlBuilder()
+        upload_config = _UploadConfig()
+
+        # Base URL may change once we know simple vs. resumable.
+        base_url = connection.API_BASE_URL + '/upload'
+        path = '/projects/%s/jobs' % (self._dataset.project,)
+        upload_url = connection.build_api_url(api_base_url=base_url, path=path)
+
+        # Use apitools 'Upload' facility.
+        request = Request(upload_url, 'POST', headers,
+                          body=json.dumps(metadata))
+
+        upload.configure_request(upload_config, request, url_builder)
+        query_params = url_builder.query_params
+        base_url = connection.API_BASE_URL + '/upload'
+        request.url = connection.build_api_url(api_base_url=base_url,
+                                               path=path,
+                                               query_params=query_params)
+        upload.initialize_upload(request, connection.http)
+
+        if upload.strategy == RESUMABLE_UPLOAD:
+            http_response = upload.stream_file(use_chunks=True)
+        else:
+            http_response = make_api_request(connection.http, request,
+                                             retries=num_retries)
+        response_content = http_response.content
+        if not isinstance(response_content,
+                          six.string_types):  # pragma: NO COVER  Python3
+            response_content = response_content.decode('utf-8')
+        return client.job_from_resource(json.loads(response_content))
+
+
+def _configure_job_metadata(metadata,  # pylint: disable=R0913
+                            allow_jagged_rows,
+                            allow_quoted_newlines,
+                            create_disposition,
+                            encoding,
+                            field_delimiter,
+                            ignore_unknown_values,
+                            max_bad_records,
+                            quote_character,
+                            skip_leading_rows,
+                            write_disposition):
+    """Helper for :meth:`Table.upload_from_file`."""
+    load_config = metadata['configuration']['load']
+
+    if allow_jagged_rows is not None:
+        load_config['allowJaggedRows'] = allow_jagged_rows
+
+    if allow_quoted_newlines is not None:
+        load_config['allowQuotedNewlines'] = allow_quoted_newlines
+
+    if create_disposition is not None:
+        load_config['createDisposition'] = create_disposition
+
+    if encoding is not None:
+        load_config['encoding'] = encoding
+
+    if field_delimiter is not None:
+        load_config['fieldDelimiter'] = field_delimiter
```

Comment:
RE: `kwargs` and remapping to camelCased, we could have a global to do that mapping for us:

```python
transformed_kw = {_REMAP[key]: value for key, value in kwargs.iteritems}
```

(Yay! No `dict` comprehensions in Py2.6, so lame.)